File: 백준/Gold/16236. 아기 상어/아기 상어.py
n= int(input())

graph = [list(map(int,input().split())) for _ in range(n)]
from collections import deque
for i in range(n):
    for j in range(n):
        if graph[i][j] == 9:
            shark = [i,j]
            graph[i][j] = 0
direction = [(0,1),(0,-1),(1,0),(-1,0)]
import sys
def search(shark):
    visit = [[0]* n for _ in range(n)]

    queue = deque([shark+[0]])
    y,x= shark
    visit[y][x] = 1
    fish = []
    fish_dist =10000
    while queue:
        y,x,dist = queue.popleft()
        if dist == fish_dist: break

        for dy,dx in direction:
            ny,nx = dy+y,dx+x
            if 0<=ny< n and 0<=nx <n and visit[ny][nx] == 0:

                if graph[ny][nx] >= 1 and graph[ny][nx] < size:
                    fish_dist = dist+1
                    fish.append([ny,nx,fish_dist])
                if graph[ny][nx] <=size:
                    queue.append([ny,nx,dist+1])
                visit[ny][nx] = 1
    fish.sort()
    if fish:
        return fish[0]
    else:
        print(sec)
        sys.exit(0)
size = 2
sec= 0
eat_fish = 0
while True:

    ny,nx,dist = search(shark)
    graph[ny][nx] = 0
    eat_fish += 1
    sec += dist
    shark = [ny,nx]

    if eat_fish == size:
        eat_fish = 0
        size +=1

File: 백준/Gold/16236. 아기 상어/아기 상어.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int n, size = 2, seconds = 0;
    private static int[][] graph;
    private static int[] shark;
    private static final int[][] direction = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    private static int eatFish = 0;
    private static Comparator<int[]> comparator = (ints1, t1) -> {
        if (ints1[0] != t1[0]) {
            return Integer.compare(ints1[0], t1[0]);
        } else {
            return Integer.compare(ints1[1], t1[1]);
        }
    };


    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        graph = new int[n][];
        for (int i = 0; i < n; i++) {
            graph[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (graph[i][j] == 9) {
                    shark = new int[]{i, j, 0};
                    graph[i][j] = 0;
                }
            }
        }
        while (true) {
            int[] solve = solve(shark);
            int ny = solve[0];
            int nx = solve[1];
            int dist = solve[2];

            graph[ny][nx] = 0;
            eatFish += 1;
            seconds += dist;
            shark = new int[]{ny, nx, 0};
            if (eatFish == size) {
                eatFish = 0;
                size += 1;
            }

        }

    }

    private static int[] solve(int[] shark) {
        int[][] visit = new int[n][n];

        ArrayDeque<int[]> queue = new ArrayDeque<>();
        queue.add(shark);
        ArrayList<int[]> fish = new ArrayList<>();
        int fishDist = Integer.MAX_VALUE;
        int[] ints = new int[0];
        while (!queue.isEmpty()) {
            ints = queue.pollFirst();
            int y = ints[0];
            int x = ints[1];
            visit[y][x] = 1;
            int dist = ints[2];
            if (dist == fishDist) {
                break;
            }
            for (int[] dir : direction) {
                int ny = y + dir[0];
                int nx = x + dir[1];

                if (0 <= ny && ny < n && 0 <= nx && nx < n && visit[ny][nx] == 0) {
                    if (graph[ny][nx] >= 1 && graph[ny][nx] < size) {
                        fishDist = dist + 1;
                        fish.add(new int[]{ny, nx, fishDist});
                    }
                    if (graph[ny][nx] <= size) {
                        queue.add(new int[]{ny, nx, dist + 1});
                    }
                    visit[ny][nx] = 1;
                }
            }
        }
        Collections.sort(fish, comparator);

        if (!fish.isEmpty()) {
            return fish.get(0);
        } else {
            System.out.println(seconds);
            System.exit(0);
        }
        return null;
    }
}
